### PR TITLE
[TT-8770] Custom policies via session metadata

### PR DESCRIPTION
**Pull request by buger**
## Description

If session metadata contains "policies" field, read JSON out of this value.

## Related Issue

https://tyktech.atlassian.net/browse/TT-8770

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/4992/merge
Commit:  
Triggered by: pull_request (@buger)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/4894679385)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/4992/merge
Commit:  
Triggered by: pull_request (@buger)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/4894679385)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4992/merge
Commit: 222c8d8d6fa28a37be107dc821fbb9bb7aadba31 
Triggered by: pull_request (@buger)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/4895198057)

File: gateway/middleware.go [hunk 1]
```diff
@@ -311,17 +312,43 @@ func (t BaseMiddleware) ApplyPolicies(session *user.SessionState) error {
...
-
-	for _, polID := range policies {
-		t.Gw.policiesMu.RLock()
-		policy, ok := t.Gw.policiesByID[polID]
-		t.Gw.policiesMu.RUnlock()
-		if !ok {
-			err := fmt.Errorf("policy not found: %q", polID)
-			t.Logger().Error(err)
-			return err
+	policyIDs := session.PolicyIDs()
```
Comment (titpetric) [inline gateway/middleware.go:315]: Could L315-331 land in a `func (s *SessionData) PolicyIDs() ([]string, error)` please. This enables protections for concurrent map access over the session.MetaData map, e.g. can implement locking/mutex. This behaviour should be local to *SessionData.

Similarly, the second block should be in a function akin to `func (gw *Gateway) getPolicies([]string) ([]user.Policy, error)`.

Inlined code is hard to test, separate functions are better.

**Review by titpetric — CHANGES_REQUESTED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4992/merge
Commit: 69dae9d96d50bdfe86b106d158788a975777a7db 
Triggered by: pull_request (@lghiur)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5021653293)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4992/merge
Commit: a1b7d82df603d97324bec8b3a788bdd6336cf4db 
Triggered by: pull_request (@MFCaballero)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5291850509)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4992/merge
Commit: 69dae9d96d50bdfe86b106d158788a975777a7db 
Triggered by: pull_request (@MFCaballero)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5291868000)

File: user/session.go [hunk 2]
```diff
@@ -329,6 +330,27 @@ func (s *SessionState) GetQuotaLimitByAPIID(apiID string) (int64, int64, int64,
 	return s.QuotaMax, s.QuotaRemaining, s.QuotaRenewalRate, s.QuotaRenews
 }
 
+func (s *SessionState) CustomPolicies() ([]Policy, error) {
+	var customPolicies []Policy
+
+	metadataPolicies, found := s.MetaData["policies"].([]interface{})
+	if !found {
+		return []Policy{}, fmt.Errorf("policies not found in metadata")
```
[the remark this answers is not in the document]
Comment (titpetric) [reply user/session.go:338]: also, use errors.New (fmt.Errorf has no formatting directives)

[the remark this answers is not in the document]
Comment (titpetric) [reply user/session.go:338]: also, this needs a `ErrNotFound` type, error handling outside of this function is missing

File: gateway/middleware.go [hunk 3]
```diff
@@ -381,21 +380,46 @@ func (t BaseMiddleware) ApplyPolicies(session *user.SessionState) error {
...
+		}
+	}
 
-			return err
+	for _, polID := range policyIDs {
+		var policy user.Policy
+
+		if useCustomPolicies {
+			polIdx, _ := strconv.ParseInt(polID, 10, 64)
+			policy = customPolicies[polIdx]
```
Comment (titpetric) [inline gateway/middleware.go:408]: Unhandled err above puts into question if `customPolicies` index is out of range.

File: gateway/policy_test.go [hunk 4]
```diff
@@ -900,22 +901,54 @@ func TestApplyPolicies(t *testing.T) {
...
 			sess := tc.session
 			if sess == nil {
 				sess = &user.SessionState{}
 			}
-			sess.SetPolicies(tc.policies...)
+
+			if useMetadata {
+				pols := []user.Policy{}
+				for _, p := range tc.policies {
+					pols = append(pols, ts.Gw.policiesByID[p])
```
[the remark this answers is not in the document]
Comment (padiazg) [reply gateway/policy_test.go:927]: done

File: gateway/middleware.go (hunk 3 above)
Comment (padiazg) [reply gateway/middleware.go:408]: error handled

Comment (titpetric) [reply gateway/middleware.go:408]: @padiazg i think you missed a push there (expected a length check or a return on L392)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4992/merge
Commit: 1550638810fafa54793c033b351d7994463bf2b2 
Triggered by: pull_request (@padiazg)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5824886177)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4992/merge
Commit: 07d5d68765d9bafd3dfe9f3114e3bccd6c80ec6f 
Triggered by: pull_request (@padiazg)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5824992832)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4992/merge
Commit: 1d9f2f56dacdbdbe9d75ee29100cd1b5512831ed 
Triggered by: pull_request (@padiazg)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5825039074)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4992/merge
Commit: cc45326b8d33fede6d16359ca2af9a45c7610ef2 
Triggered by: pull_request (@padiazg)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5834300244)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4992/merge
Commit: 5d4ded9f08a360157de3720873022290cd7b104b 
Triggered by: pull_request (@padiazg)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5834895925)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4992/merge
Commit: 1b9dfe6fe884567667ef66623a017b57911a4e16 
Triggered by: pull_request (@padiazg)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5834952178)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4992/merge
Commit: 9c2668090cfcb6a8035e835899e6db55285b8f86 
Triggered by: pull_request (@padiazg)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5835006041)

**Review by titpetric — APPROVED** [review]
@buger this LGTM, the middleware change is covered by existing tests (TestApplyPolicies).

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/4992/merge
Commit: 1e635cac862160e413d0bc24b05ad46a176e45f0 
Triggered by: pull_request (@padiazg)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5835975987)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4992/merge
Commit: 3d8d1710a145887cef743ac93c726915615297a1 
Triggered by: pull_request (@padiazg)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5835995738)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/4992/merge
Commit: f12f1201dc768d7b2fbf57e3c1d7c5096e16742c 
Triggered by: pull_request (@padiazg)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5836252181)

**Pull request merged**